internal/saucecloud/imagerunner.go

```diff
@@ -0,0 +1,361 @@
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type execResult struct {
+	name      string
+	runID     string
+	status    string
+	err       error
+	duration  time.Duration
+	startTime time.Time
+	endTime   time.Time
+	attempts  int
+}
+
+func (r *ImgRunner) RunProject() (int, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	r.ctx = ctx
+	r.cancel = cancel
+
+	sigChan := r.registerInterruptOnSignal()
+	defer unregisterSignalCapture(sigChan)
+
+	suites, results := r.createWorkerPool(1, 0)
+
+	// Submit suites to work on.
+	go func() {
+		for _, s := range r.Project.Suites {
+			suites <- s
+		}
+	}()
+
+	if passed := r.collectResults(results, len(r.Project.Suites)); !passed {
+		return 1, nil
+	}
+
+	return 0, nil
+}
+
+func (r *ImgRunner) createWorkerPool(ccy int, maxRetries int) (chan imagerunner.Suite, chan execResult) {
+	suites := make(chan imagerunner.Suite, maxRetries+1)
+	results := make(chan execResult, ccy)
+
+	log.Info().Int("concurrency", ccy).Msg("Launching workers.")
+	for i := 0; i < ccy; i++ {
+		go r.runSuites(suites, results)
+	}
+
+	return suites, results
+}
+
+func (r *ImgRunner) runSuites(suites chan imagerunner.Suite, results chan<- execResult) {
+	for suite := range suites {
+		// Apply defaults.
+		defaults := r.Project.Defaults
+		if defaults.Name != "" {
+			suite.Name = defaults.Name + " " + suite.Name
+		}
+
+		suite.Image = orDefault(suite.Image, defaults.Image)
+		suite.ImagePullAuth = orDefault(suite.ImagePullAuth, defaults.ImagePullAuth)
+		suite.EntryPoint = orDefault(suite.EntryPoint, defaults.EntryPoint)
+		suite.Timeout = orDefault(suite.Timeout, defaults.Timeout)
+		suite.Files = append(suite.Files, defaults.Files...)
+		suite.Artifacts = append(suite.Artifacts, defaults.Artifacts...)
+
+		if suite.Env == nil {
+			suite.Env = make(map[string]string)
+		}
+		for k, v := range defaults.Env {
+			suite.Env[k] = v
+		}
+
+		startTime := time.Now()
+
+		if r.ctx.Err() != nil {
+			results <- execResult{
+				name:      suite.Name,
+				startTime: startTime,
+				endTime:   time.Now(),
+				duration:  time.Since(startTime),
+				status:    imagerunner.StateCancelled,
+				err:       ErrSuiteCancelled,
+			}
+			continue
+		}
+
+		run, err := r.runSuite(suite)
+
+		results <- execResult{
+			name:      suite.Name,
+			runID:     run.ID,
+			status:    run.Status,
+			err:       err,
+			startTime: startTime,
+			endTime:   time.Now(),
+			duration:  time.Since(startTime),
+			attempts:  1,
+		}
+	}
+}
+
+func (r *ImgRunner) runSuite(suite imagerunner.Suite) (imagerunner.Runner, error) {
+	var run imagerunner.Runner
+	metadata := make(map[string]string)
+	metadata["name"] = suite.Name
+
+	files, err := mapFiles(suite.Files)
+	if err != nil {
+		log.Err(err).Str("suite", suite.Name).Msg("Unable to read source files")
+		return run, err
+	}
+
+	log.Info().Str("image", suite.Image).Str("suite", suite.Name).Msg("Starting suite.")
+
+	if suite.Timeout <= 0 {
+		suite.Timeout = 24 * time.Hour
```

Comment:
That looks a long timeout !

Reply:
It's what we have by default in other places, effectively meaning "no timeout". The platform will time a job out long before that (usually within 4h).